pootle/apps/pootle_store/views.py

```diff
@@ -350,63 +366,105 @@ def get_units(request):
+                else:
+                    max_field = sort_by
+                    sort_by = 'sort_by_field'
+                # It's necessary to use `Max()` here because we can't
+                # use `distinct()` and `order_by()` at the same time
+                units_qs = units_qs.annotate(sort_by_field=Max(max_field))
+            units_qs = units_qs.order_by(
+                sort_by, "store__pootle_path", "index")
+
+    if all(x is not None for x in [search, sfields]):
+        # Accept `sfields` to be a comma-separated string of fields (#46)
+        GET = request.GET.copy()
+        if isinstance(sfields, unicode) and u',' in sfields:
+            GET.setlist('sfields', sfields.split(u','))
+
+        # use the search form for validation only
+        search_form = make_search_form(GET)
+
+        if search_form.is_valid():
+            exact = 'exact' in search_form.cleaned_data['soptions']
+            text = search_form.cleaned_data['search']
+            sfields = GET.getlist("sfields")
+            units_qs = UnitTextSearch(
+                units_qs).search(text, sfields, exact=exact)
 
-    is_initial_request = request.GET.get('initial', False)
-    chunk_size = request.GET.get('count', limit)
-    uids_param = filter(None, request.GET.get('uids', '').split(u','))
     uids = filter(None, map(to_int, uids_param))
 
-    units = []
     unit_groups = []
-    uid_list = []
-
+    uid_list = None
+    begin = 0
+    end = 2 * chunk_size
     if is_initial_request:
-        sort_by_field = None
-        if len(step_queryset.query.order_by) > 2:
-            sort_by_field = step_queryset.query.order_by[0]
-
-        sort_on = None
-        for key, item in ALLOWED_SORTS.items():
-            if sort_by_field in item.values():
-                sort_on = key
-                break
-
-        if sort_by_field is None or sort_on == 'units':
-            uid_list = list(step_queryset.values_list('id', flat=True))
-        else:
-            uid_list = [u['id'] for u
-                        in step_queryset.values('id', 'sort_by_field')]
+        uid_list = list(units_qs.values_list('id', flat=True))
         if len(uids) == 1:
-            try:
-                uid = uids[0]
-                index = uid_list.index(uid)
-                begin = max(index - chunk_size, 0)
-                end = min(index + chunk_size + 1, len(uid_list))
-                uids = uid_list[begin:end]
-                units = step_queryset[begin:end]
-            except ValueError:
-                raise Http404  # `uid` not found in `uid_list`
-        else:
-            count = 2 * chunk_size
-            uids = uid_list[:count]
-
-    if not units and uids:
-        units = step_queryset.filter(id__in=uids)
-
-    units_by_path = groupby(units, lambda x: x.store.pootle_path)
+            if uids[0] not in uid_list:
+                raise Http404
+            uid = uids[0]
+            index = uid_list.index(uid)
```

Comment:
I don't particularly like creating unnecessary variables, so what about replacing this two lines with `index = uid_list.index(uids[0])`?